Reset algorithm is incorrect for select element with first option disabled

Ahmad-S792 opened the pull request:
#### 5aa94a8e95f6432c20524bae9b665148b454e6c9
<pre>
Reset algorithm is incorrect for select element with first option disabled

Reset algorithm is incorrect for select element with first option disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=249849">https://bugs.webkit.org/show_bug.cgi?id=249849</a>

Reviewed by Tim Nguyen.

This patch is to align WebKit with Blink / Chromium, Gecko / Firefox and Web-Specification.

Web-Spec - <a href="https://www.w3.org/TR/html5/forms.html#ask-for-a-reset">https://www.w3.org/TR/html5/forms.html#ask-for-a-reset</a>

Partial Merge - <a href="https://chromium.googlesource.com/chromium/src.git/+/5b2038755e5c1e046642373d6b3add2452b8eb2e">https://chromium.googlesource.com/chromium/src.git/+/5b2038755e5c1e046642373d6b3add2452b8eb2e</a>

As per the spec, the user agent must set the selectedness of the first option element in the list
of options in tree order that is not disabled, if any, to true.

* Source/WebCore/html/HTMLSelectElement.cpp:
(HTMLSelectElement::reset): Update &apos;if&apos; to also disallow selecting disabled option
* LayoutTests/fast/forms/select-ask-for-reset.html: Add Test Case
* LayoutTests/fast/forms/select-ask-for-reset-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/258313@main">https://commits.webkit.org/258313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90f5f85c1353c15e60cbb0b8e5811420f2acb5f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101488 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10644 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110765 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171015 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105468 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1541 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93881 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108576 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92066 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35355 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90720 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23476 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78353 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4244 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24985 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4309 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1442 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44473 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5708 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6071 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/forms/select-ask-for-reset.html:
```diff
@@ -0,0 +1,31 @@
+<script src="../../resources/testharnessreport.js"></script>
+<link rel="help" href="https://html.spec.whatwg.org/multipage/forms.html#ask-for-a-reset">
+<div id="log"></div>
+<script>
```
Some of these commits: https://github.com/chromium/chromium/commits/77578ccb4082ae20a9326d9e673225f1189ebb63/third_party/WebKit/LayoutTests/fast/forms/select/select-ask-for-reset.html?browsing_rename_history=true&new_path=third_party/blink/web_tests/fast/forms/select/select-ask-for-reset.html&original_branch=main 

might make the subtest that you removed pass.

Ahmad-S792 replied:
Yes - I am planning to work on them one by one. It is just my skills level restricting me to take big patches.

nt1m commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1091,7 +1091,7 @@ void HTMLSelectElement::reset()
             option.setSelectedState(false);
 
-        if (!firstOption)
+        if (!firstOption &&  !option->isDisabledFormControl())
```
```suggestion
        if (!firstOption && !option.isDisabledFormControl())
```

Ahmad-S792 replied:
Done

nt1m reviewed (approved):
r=me if EWS is green

Ahmad-S792 committed:
Reset algorithm is incorrect for select element with first option disabled

Reset algorithm is incorrect for select element with first option disabled
https://bugs.webkit.org/show_bug.cgi?id=249849

Reviewed by Tim Nguyen.

This patch is to align WebKit with Blink / Chromium, Gecko / Firefox and Web-Specification.

Web-Spec - https://www.w3.org/TR/html5/forms.html#ask-for-a-reset

Partial Merge - https://chromium.googlesource.com/chromium/src.git/+/5b2038755e5c1e046642373d6b3add2452b8eb2e

As per the spec, the user agent must set the selectedness of the first option element in the list
of options in tree order that is not disabled, if any, to true.

* Source/WebCore/html/HTMLSelectElement.cpp:
(HTMLSelectElement::reset): Update 'if' to also disallow selecting disabled option
* LayoutTests/fast/forms/select-ask-for-reset.html: Add Test Case
* LayoutTests/fast/forms/select-ask-for-reset-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/258313@main
